// ng-swipe-card/esm5/lib/card-wrapper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, ContentChildren, QueryList } from '@angular/core';
import { SwipeableDirective } from './swipeable.directive';
var CardWrapperComponent = /** @class */ (function () {
    function CardWrapperComponent() {
    }
    /**
     * @return {?}
     */
    CardWrapperComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    CardWrapperComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    CardWrapperComponent.prototype.like = /**
     * @return {?}
     */
    function () {
        if (this.cards.first) {
            this.cards.first.simulateSwipe(true);
        }
    };
    /**
     * @return {?}
     */
    CardWrapperComponent.prototype.dislike = /**
     * @return {?}
     */
    function () {
        if (this.cards.first) {
            this.cards.first.simulateSwipe(false);
        }
    };
    CardWrapperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sw-card-wrapper',
                    template: "\n    <ng-content></ng-content>",
                    styles: [":host {\n    position: relative;\n    display: block;\n  }"]
                },] },
    ];
    CardWrapperComponent.ctorParameters = function () { return []; };
    CardWrapperComponent.propDecorators = {
        cards: [{ type: ContentChildren, args: [SwipeableDirective,] }]
    };
    return CardWrapperComponent;
}());
export { CardWrapperComponent };
if (false) {
    /** @type {?} */
    CardWrapperComponent.prototype.cards;
}
