// ng-swipe-card/esm2015/lib/card-wrapper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, ContentChildren, QueryList } from '@angular/core';
import { SwipeableDirective } from './swipeable.directive';
export class CardWrapperComponent {
    constructor() {
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
    }
    /**
     * @return {?}
     */
    like() {
        if (this.cards.first) {
            this.cards.first.simulateSwipe(true);
        }
    }
    /**
     * @return {?}
     */
    dislike() {
        if (this.cards.first) {
            this.cards.first.simulateSwipe(false);
        }
    }
}
CardWrapperComponent.decorators = [
    { type: Component, args: [{
                selector: 'sw-card-wrapper',
                template: `
    <ng-content></ng-content>`,
                styles: [`:host {
    position: relative;
    display: block;
  }`]
            },] },
];
CardWrapperComponent.ctorParameters = () => [];
CardWrapperComponent.propDecorators = {
    cards: [{ type: ContentChildren, args: [SwipeableDirective,] }]
};
if (false) {
    /** @type {?} */
    CardWrapperComponent.prototype.cards;
}
